packages/MapFactory/src/DTK_C_API_Map.hpp

```diff
@@ -0,0 +1,174 @@
+/****************************************************************************
+ * Copyright (c) 2012-2018 by the DataTransferKit authors                   *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the DataTransferKit library. DataTransferKit is     *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+/*!
+ * \file
+ * \brief C adapter to DTK Maps.
+ */
+#ifndef DTK_C_API_MAP_HPP
+#define DTK_C_API_MAP_HPP
+
+#include <memory>
+
+#include <DataTransferKit_config.hpp>
+#include <DTK_C_API.h>
+#include <DTK_C_API.hpp>
+#include <DTK_ParallelTraits.hpp>
+#include <DTK_UserApplication.hpp>
+#include <DTK_NearestNeighborOperator.hpp>
+
+#include <Teuchos_DefaultMpiComm.hpp>
+
+#include <mpi.h>
+
+#include <memory>
+#include <string>
+
+namespace DataTransferKit
+{
+//---------------------------------------------------------------------------//
+struct DTK_Map
+{
+    virtual ~DTK_Map() = default;
+
+    virtual void apply( const std::string& source_field_name,
+                        const std::string& target_field_name ) = 0;
+};
+
+//---------------------------------------------------------------------------//
+template<class Device>
+struct DTK_MapImpl : public DTK_Map
+{
+    DTK_MapImpl( MPI_Comm comm,
+                 DTK_UserApplicationHandle source,
+                 DTK_UserApplicationHandle target )
+        : _source( reinterpret_cast<DTK_Registry*>(source)->_registry )
+        , _target( reinterpret_cast<DTK_Registry*>(target)->_registry )
+    {
+        // Create a Teuchos comm from the MPI comm.
+        auto teuchos_comm = Teuchos::rcp( new Teuchos::MpiComm<int>(comm) );
+
+        // FOR NOW JUST CREATE A NEAREST NEIGHBOR OPERATOR FOR DEMONSTRATION
+        // PURPOSES. THIS WILL BE REPLACED BY A PROPER FACTORY.
```

Comment:
You planning on adding arguments to the constructor to specify which map to build?
Have you given much thoughts on what these might be?

Reply:
For now I'm hoping that a string of parameters is sufficient to build the maps. A map should know which user data functions it needs to populate its data structures and if those functions aren't implemented a user should get clear error messages.